ff/serving/coordinator/coordinator.go

```diff
@@ -0,0 +1,271 @@
+		return err
+	}
+	providerEntry, err := ts.FetchProvider(c.Metadata, context.Background())
+	if err != nil {
+		return err
+	}
+	p, err := provider.Get(provider.Type(providerEntry.Type()), providerEntry.SerializedConfig())
+	if err != nil {
+		return err
+	}
+	store, err := p.AsOfflineStore()
+	if err != nil {
+		return err
+	}
+	providerResID := provider.ResourceID{Name: resID.Name, Variant: resID.Variant}
+	if _, err := store.GetTrainingSet(providerResID); err == nil {
+		return fmt.Errorf("Training set already exists")
+	}
+	features := ts.Features()
+	featureList := make([]provider.ResourceID, len(features))
+	for i, feature := range features {
+		featureList[i] = provider.ResourceID{Name: feature.Name, Variant: feature.Variant}
+	}
+	label, err := ts.FetchLabel()
+	if err != nil {
+		return err
+	}
+	trainingSetDef := provider.TrainingSetDef{
+		ID:       providerResID,
+		Label:    provider.ResourceID{Name: label.GetName(), Variant: label.GetVariant()},
+		Features: featureList,
+	}
+	tsRunnerConfig := runner.TrainingSetRunnerConfig{
+		OfflineType:   providerEntry.GetType(),
+		OfflineConfig: providerEntry.GetSerializedConfig(),
+		Def:           trainingSetDef,
+	}
+	jobRunner, err := c.GetJobRunner("CREATE_TRAINING_SET", tsRunnerConfig)
+	if err != nil {
+		return err
+	}
+	completionWatcher, err := jobRunner.Run()
+	if err != nil {
+		return err
+	}
+	if err := completionWatcher.Wait(); err != nil {
+		return err
+	}
+	if err := c.Metadata.SetStatus(context.Background(), resID, metadata.ResourceStatus.Ready); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Coordinator) safeGet(mtx *concurrency.Mutex, key string) ([]byte, error) {
+	keyValueClient := clientv3.NewKV(c.ETCDClient)
+	txn := (keyValueClient).Txn(context.Background())
+	response, err := txn.If(mtx.IsOwner()).Then(clientv3.OpGet(jobKey)).Commit()
+	if err != nil {
+		return nil, fmt.Errorf("transaction did not succeed: %v", err)
+	}
+	if !response.Succeeded {
+		return nil, fmt.Errorf("transaction did not succeed")
+	}
+	var responseData []byte
+	if len(response.Responses[0].GetResponseRange().GetKvs()) > 0 {
+		responseData = response.Responses[0].GetResponseRange().GetKvs()[0].Value
+	} else {
+		return nil, fmt.Errorf("Coordinator job does not exist. %v")
+	}
+	return responseData, nil
+
+}
+
+func (c *Coordinator) incrementJobInETCD(mtx *concurrency.Mutex, job CoordinatorJob, jobKey string) error {
+	job.Attempts += 1
+	serializedJob, err := job.Serialize()
+	if err != nil {
+		return fmt.Errorf("Could not serialize coordinator job. %v", err)
+	}
+	keyValueClient := clientv3.NewKV(c.ETCDClient)
+	txn := (keyValueClient).Txn(context.Background())
+	if _, err := txn.If(l.IsOwner()).Then(clientv3.OpPut(jobKey, serialized)).Commit(); err != nil {
+		return fmt.Errorf("Could not set iterated coordinator job. %v", err)
+	}
+	return nil
+}
+
+func (c *Coordinator) safeDelete(mtx *concurrency.Mutex, key string) error {
+	keyValueClient := clientv3.NewKV(c.ETCDClient)
```

Comment:
why not just create this once when you create the coordinator and make it a field

Reply:
rename to kvClient? or kvInterface, even etcdKVInterface